zio-http/src/main/scala/zio/http/Handler.scala

```diff
@@ -0,0 +1,632 @@
+package zio.http
+
+import io.netty.handler.codec.http.HttpHeaderNames
+import zio._
+import zio.http.html.{Html, Template}
+import zio.http.model._
+import zio.http.model.headers.HeaderModifierZIO
+import zio.http.socket.{SocketApp, WebSocketChannelEvent}
+import zio.stream.ZStream
+
+import java.io.{File, FileNotFoundException}
+import java.nio.charset.Charset
+import java.nio.file.Paths
+import java.util.zip.ZipFile
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+sealed trait OptionalHandler[-R, +Err, -In, +Out]
```

Comment:
Looking at this it made me wonder, why have the `sealed trait` here instead of `Route`?

If `Route` is a sealed trait, then every route is either a pattern and a handler, or the empty route, which has neither pattern nor handler.

(I know this was my suggestion but it seems cleaner to push the sum type to the route.)

Thoughts?

Reply:
Is the empty case a route that doesn't want to handle anything?

The other possibility this opens that I don't think the current model does is you could have a route which as a result of its input returns another route. Basically `Http` becomes a "tree" that eventually either chooses not to handle something or has leaves with handlers. Don't think that is necessary for anything currently but it is kind of interesting and opens up another mode of composition.

Reply:
Actually I think this change makes my "middleware semantics" invalid, or at lest harder to understand - that you can apply a "handler middleware" to a route and it is just get applied "immediately on the next level" without affecting the routing. I was trying to keep the composition on the route level to minimum (to just have orElse, basically) to keep it simple and to solve the original issue of weird behavior of middlewares.

So even though it may make it possible to express more interesting composed routes let's think it through how it relates to applying middlewares and not running into surprises.

Reply:
💯

Reply:
I think that `Http` (`Route`) is basically a routing table, something like:

| Http Method | Path                 | Handler               |
| ------------ | -------------------- | --------------------- |
| GET         | /users                | getAllUsers            |
| GET         | /users/{id}           | getUserById            |
| POST        | /users                | createNewUser           |
| PUT         | /users/{id}           | updateUserById          |
| DELETE      | /users/{id}           | deleteUserById         |
| GET         | /orders                | getAllOrders            |
| GET         | /orders/{id}           | getOrderById            |
| POST        | /orders                | createNewOrder           |
| PUT         | /orders/{id}           | updateOrderById          |
| DELETE      | /orders/{id}           | deleteOrderById         |

Why can't a table be empty? 

Right now we fake emptiness by saying it's a route that returns a handler that doesn't want to handle its input (null or Unhandled). But that seems equivalent to me (but worse than) just supporting an empty route.

A "tree" almost suggests to me a way to unify this and the endpoint stuff. But doing that while preserving the current API is not possible. 

It's worth discussing this a bit more as I think `OptionalHandler` is really the only unsightly bit in this new model.

Reply:
👍